crates/proof_of_stake/src/lib.rs

```diff
@@ -2913,3 +2948,110 @@ where
+        let (start, end) = delegations.last_range;
+        if let Some(end) = end {
+            // Add the `last_range` pair to the `prev_ranges` and make a new
+            // `last_range`
+            if epoch == end {
+                // This case would occur if in the same epoch, the bond was
+                // fully unbonded, followed by the bonding of new tokens
+                delegations.last_range.1 = None;
+            } else {
+                delegations.prev_ranges.insert(start, end);
+                delegations.last_range = (epoch, None);
+            }
+            bond_holders.insert(
+                storage,
+                validator.clone(),
+                delegations.clone(),
+            )?;
+        } else {
+            // do nothing since the last bond is still active
+        }
+    } else {
+        // Make a new delegation to this source-validator pair
+        let first_delegation = DelegationEpochs {
+            prev_ranges: BTreeMap::new(),
+            last_range: (epoch, None),
+        };
+        bond_holders.insert(storage, validator.clone(), first_delegation)?;
+    }
+
+    // Todo: possibly update the data by pruning old pairs in
+    // `delegations.prev_ranges`??
+
+    Ok(())
+}
+
+fn remove_delegation_target<S>(
+    storage: &mut S,
+    params: &PosParams,
+    delegator: &Address,
+    validator: &Address,
+    epoch: Epoch,
+    current_epoch: Epoch,
+) -> namada_storage::Result<()>
+where
+    S: StorageRead + StorageWrite,
+{
+    let validators = delegation_targets_handle(delegator);
+    if let Some(delegation) = validators.get(storage, validator)?.as_mut() {
+        let (_start, end) = &mut delegation.last_range;
+        debug_assert!(
+            end.is_none(),
+            "End epoch should be None since we are removing the delegation
+              right now!!"
+        );
+        *end = Some(epoch);
+        prune_old_delegations(params, delegation, current_epoch)?;
+        validators.insert(storage, validator.clone(), delegation.clone())?;
+    } else {
+        panic!("Delegation should exist since we are removing it right now!!!");
+    }
+
+    Ok(())
+}
+
+fn prune_old_delegations(
+    params: &PosParams,
+    delegations: &mut DelegationEpochs,
+    current_epoch: Epoch,
+) -> namada_storage::Result<()> {
+    let delta =
+        crate::epoched::OffsetMaxProposalPeriodOrSlashProcessingLenPlus::value(
+            params,
+        );
+    let oldest_to_keep = current_epoch.checked_sub(delta).unwrap_or_default();
+
+    let to_remove = delegations
+        .prev_ranges
+        .iter()
+        .filter_map(|(start, end)| {
+            if start < &oldest_to_keep && end < &oldest_to_keep {
+                Some(start)
+            } else {
+                None
+            }
+        })
+        .cloned()
+        .collect::<BTreeSet<_>>();
+
+    for start_epoch in to_remove {
+        delegations.prev_ranges.remove(&start_epoch);
+    }
```

Comment:
```suggestion
    delegations
        .prev_ranges
        .retain(|_start, end| *end >= oldest_to_keep);
```
(we shouldn't need to check `start` as it's always <= `end`)

Reply:
thx for catching this, done in aafc6db8a3f75c90ea463a66f5a63d62d33277ce